Added deployment dependent checks during uninstallation

ezio-auditore opened the pull request:
Signed-off-by: Kaustav Majumder <kmajumde@redhat.com>

bindrad commented on controllers/managedocs_controller.go:
```diff
@@ -1669,6 +1680,16 @@ func (r *ManagedOCSReconciler) findOCSVolumeClaims() (bool, error) {
 
+func (r *ManagedOCSReconciler) findStorageConsumerList() (bool, error) {
+	storageConsumerList := ocsv1alpha1.StorageConsumerList{}
+	if err := r.UnrestrictedClient.List(r.ctx, &storageConsumerList); err != nil {
```
Why do we need unrestrictedClient to list storageConsumer? All the storageConsumer will be in openshift-storage namespace, right?

ezio-auditore replied:
Yes correct. Made the change

bindrad commented on controllers/managedocs_controller.go:
```diff
@@ -575,15 +576,25 @@ func (r *ManagedOCSReconciler) reconcilePhases() (reconcile.Result, error) {
+					return ctrl.Result{}, err
+				}
+				if found {
+					r.Log.Info("Found consumer PVCs using OCS storageclasses, cannot proceed on uninstallation")
```
```suggestion
					r.Log.Info("Found consumer PVs using OCS storageclasses, cannot proceed on uninstallation")
```

ezio-auditore replied:
Done

nb-ohad commented on controllers/managedocs_controller.go:
```diff
@@ -1669,6 +1680,16 @@ func (r *ManagedOCSReconciler) findOCSVolumeClaims() (bool, error) {
 	return false, nil
 }
 
+func (r *ManagedOCSReconciler) findStorageConsumerList() (bool, error) {
```
This function returns a bool so the name is a little bit misleading. Can we please rename it to `hasOCSStorageConsumers`? In in the same manner rename `findOCSVolumeClaims` to `hasOCSVolumes`? 

Please note that `findOCSVolumeClaims` is not searching for volume claims, it searches for volumes, so the current name is also problematic in that regard.

Why the need to mention List in the function name? can we follow the `findOCSVolumeClaims` convention and call this `findOCSStorageConsumers`. Also `findOCSVolumeClaims` does not search for VolumeClaims, it searches for Volumes so let's also rename it to `

nb-ohad commented on controllers/managedocs_controller.go:
```diff
@@ -1669,6 +1680,16 @@ func (r *ManagedOCSReconciler) findOCSVolumeClaims() (bool, error) {
+	if err := r.list(&storageConsumerList); err != nil {
+		return false, fmt.Errorf("unable to list storage consumers: %v", err)
+	}
+	if len(storageConsumerList.Items) > 0 {
+		return true, nil
+	}
+	return false, nil
```
```suggestion
	return len(storageConsumerList.Items) > 0, nil
```

nb-ohad commented on controllers/managedocs_controller.go:
```diff
@@ -575,15 +576,25 @@ func (r *ManagedOCSReconciler) reconcilePhases() (reconcile.Result, error) {
+					return ctrl.Result{}, err
+				}
+				if found {
+					r.Log.Info("Found Storage consumers, cannot proceed on uninstallation")
```
```suggestion
					r.Log.Info("Found OCS  storage consumers, cannot proceed with uninstallation")
```

nb-ohad commented on controllers/managedocs_controller.go:
```diff
@@ -575,15 +576,25 @@ func (r *ManagedOCSReconciler) reconcilePhases() (reconcile.Result, error) {
+					return ctrl.Result{}, err
+				}
+				if found {
+					r.Log.Info("Found consumer PVs using OCS storageclasses, cannot proceed on uninstallation")
```
```suggestion
					r.Log.Info("Found consumer PVs using OCS storageclasses, cannot proceed with uninstallation")
```

nb-ohad commented:
@ezio-auditore Looks good to me, but please add tests before merging

leelavg commented on controllers/managedocs_controller.go:
```diff
@@ -1669,6 +1681,13 @@ func (r *ManagedOCSReconciler) findOCSVolumeClaims() (bool, error) {
 }
 
+func (r *ManagedOCSReconciler) hasOCSStorageConsumers() (bool, error) {
+	storageConsumerList := ocsv1alpha1.StorageConsumerList{}
```
- it seems `kind: StorageConsumer` isn't part of either `Watches`/`Owns`and I could see below issues with that:
1. straightforward, we don't get a reconcile event when resource changes
2. no cache is built for this object in operator-sdk

- if we want ocs-operator to be the sole owner I'd suggest using `r.UnrestrictedClient`
- just confirming, `storageconsumer` will always be created in deployer namespace right?

minor optimisation: since we want presence or absence why not restrict the list to only one item?
```
r.Client.List(ctx, storageConsumerList, &client.ListOptions{Limit: int64(1)})
```

ezio-auditore replied:
> * it seems `kind: StorageConsumer` isn't part of either `Watches`/`Owns`and I could see below issues with that:
> 
> 1. straightforward, we don't get a reconcile event when resource changes
Do we even need to watch this, since this should only reconcile during uninstallation
> 2. no cache is built for this object in operator-sdk
I don't think we need a cache for this. Maybe @nb-ohad can clarify
> 
> * if we want ocs-operator to be the sole owner I'd suggest using `r.UnrestrictedClient`
Why? It will be in the same namespace
> * just confirming, `storageconsumer` will always be created in deployer namespace right?
Yes
> 
> minor optimisation: since we want presence or absence why not restrict the list to only one item?
> 
> ```
> r.Client.List(ctx, storageConsumerList, &client.ListOptions{Limit: int64(1)})
> ```
Yeah! this change seems more optimised. Thanks.

nb-ohad replied:
@leelavg 
+1 for the suggestion to limit the list results to 1. 
I think that the `hasOCSVolumes` util can benefit from this enhancement as well.

Regarding the `watch` or using `unrestrictedClient` I don't really understand why should we do it. 
This code will only be invoked when we are instructed to uninstall the offering and act as a guard, in the case of condition failure we are requeuing with a very specific "after" time (10 seconds) to try again. 
This means no watch is needed as there is no point in trying to count and store a counter of creations and deletions of storage consumers throughout the lifetime of the deployment for a single check when we want to uninstall. 

To clarify, a watch is usually used in two very specific use cases:
1. Watch input resources to get changes in the desired state
2. Watch reconciled resources (output resources) to revert changes to them if they are changed by another agent and divert from the desired state

Out use case here does not match any of the two

leelavg replied:
> Out use case here does not match any of the two
- ack, my understanding might be wrong here then
- for posterity below is what I was thinking 

1. When we watch/own resources, sdk creates cache for us and thought maintaining that as well will be useful even during uninstall 
2. If there's a cache miss there might be a cool off period after which the request will be sent to api-server but not immediately

- from my comment pls implement only limiting results and pls discard other requests based on @nb-ohad reasoning

nb-ohad replied:
@leelavg 
I think you are mixing between the k8s client cache (for clients that were created with the cache option), which to the best of my understanding, has nothing to do with controller watch. 
And the way that op-sdk decides when to implement controller watch, which internally uses the k8s client.

The content of the client cache is based on the client usage (mainly `get`, and `list` operations, but also `watch`), but might also have some relation to the controller watch which is implemented using the client semantics. 

In any case, I still think my reasoning above (last comment) stands still.

ezio-auditore replied:
Done

ezio-auditore commented:
To be merged when #200 merges.

nb-ohad commented:
@ezio-auditore please attach the output of a successful `make test` as a comment. 
Similar to how it is done here: https://github.com/red-hat-storage/ocs-osd-deployer/pull/197#issuecomment-1135772387

ezio-auditore commented:
make test
which: no controller-gen in (/usr/local/go/bin:/home/kmajumde/.nvm/versions/node/v14.18.2/bin:/home/kmajumde/.local/bin:/home/kmajumde/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/var/lib/snapd/snap/bin:/home/kmajumde/bin:/bin:/usr/local/go/bin:/home/kmajumde/bin:/bin:/usr/local/go/bin:/home/kmajumde/bin:/bin:/usr/local/go/bin)
which: no kustomize in (/usr/local/go/bin:/home/kmajumde/.nvm/versions/node/v14.18.2/bin:/home/kmajumde/.local/bin:/home/kmajumde/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/var/lib/snapd/snap/bin:/home/kmajumde/bin:/bin:/usr/local/go/bin:/home/kmajumde/bin:/bin:/usr/local/go/bin:/home/kmajumde/bin:/bin:/usr/local/go/bin)
go: creating new go.mod: module tmp
go get: installing executables with 'go get' in module mode is deprecated.
        To adjust and download dependencies of the current module, use 'go get -d'.
        To install using requirements of the current module, use 'go install'.
        To install ignoring the current module, use 'go install' with a version,
        like 'go install example.com/cmd@latest'.
        For more information, see https://golang.org/doc/go-get-install-deprecation
        or run 'go help get' or 'go help install'.
go get: added github.com/fatih/color v1.7.0
go get: added github.com/gobuffalo/flect v0.2.0
go get: added github.com/gogo/protobuf v1.3.1
go get: added github.com/google/gofuzz v1.1.0
go get: added github.com/inconshreveable/mousetrap v1.0.0
go get: added github.com/json-iterator/go v1.1.8
go get: added github.com/mattn/go-colorable v0.1.2
go get: added github.com/mattn/go-isatty v0.0.8
go get: added github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go get: added github.com/modern-go/reflect2 v1.0.1
go get: added github.com/spf13/cobra v1.0.0
go get: added github.com/spf13/pflag v1.0.5
go get: added golang.org/x/mod v0.2.0
go get: added golang.org/x/net v0.0.0-20200226121028-0de0cce0169b
go get: added golang.org/x/sys v0.0.0-20191022100944-742c48ecaeb7
go get: added golang.org/x/text v0.3.2
go get: added golang.org/x/tools v0.0.0-20200616195046-dc31b401abb5
go get: added golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543
go get: added gopkg.in/inf.v0 v0.9.1
go get: added gopkg.in/yaml.v2 v2.2.8
go get: added gopkg.in/yaml.v3 v3.0.0-20190905181640-827449938966
go get: added k8s.io/api v0.18.2
go get: added k8s.io/apiextensions-apiserver v0.18.2
go get: added k8s.io/apimachinery v0.18.2
go get: added k8s.io/klog v1.0.0
go get: added k8s.io/utils v0.0.0-20200324210504-a9aa75ae1b89
go get: added sigs.k8s.io/controller-tools v0.4.1
go get: added sigs.k8s.io/structured-merge-diff/v3 v3.0.0
go get: added sigs.k8s.io/yaml v1.2.0
/home/kmajumde/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
/home/kmajumde/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
mkdir -p /home/kmajumde/work/git/ocs-osd-deployer/testbin
test -f /home/kmajumde/work/git/ocs-osd-deployer/testbin/setup-envtest.sh || curl -sSLo /home/kmajumde/work/git/ocs-osd-deployer/testbin/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.6.3/hack/setup-envtest.sh
source /home/kmajumde/work/git/ocs-osd-deployer/testbin/setup-envtest.sh; fetch_envtest_tools /home/kmajumde/work/git/ocs-osd-deployer/testbin
Using cached envtest tools from /home/kmajumde/work/git/ocs-osd-deployer/testbin
source /home/kmajumde/work/git/ocs-osd-deployer/testbin/setup-envtest.sh; setup_envtest_env /home/kmajumde/work/git/ocs-osd-deployer/testbin; \
DEPLOYMENT_TYPE=converged go test ./... -coverprofile converged-cover.out
setting up env vars
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 82.507s coverage: 75.7% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    5.356s  coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
source /home/kmajumde/work/git/ocs-osd-deployer/testbin/setup-envtest.sh; setup_envtest_env /home/kmajumde/work/git/ocs-osd-deployer/testbin; \
DEPLOYMENT_TYPE=provider go test ./... -coverprofile provider-cover.out
setting up env vars
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 84.072s coverage: 73.7% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    5.424s  coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
source /home/kmajumde/work/git/ocs-osd-deployer/testbin/setup-envtest.sh; setup_envtest_env /home/kmajumde/work/git/ocs-osd-deployer/testbin; \
DEPLOYMENT_TYPE=consumer go test ./... -coverprofile consumer-cover.out
setting up env vars
?       github.com/red-hat-storage/ocs-osd-deployer     [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/api/v1alpha1        [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/controllers 71.011s coverage: 72.5% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/readinessProbe      [no test files]
ok      github.com/red-hat-storage/ocs-osd-deployer/readinessProbe/readiness    6.403s  coverage: 73.3% of statements
?       github.com/red-hat-storage/ocs-osd-deployer/templates   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/testutils   [no test files]
?       github.com/red-hat-storage/ocs-osd-deployer/utils       [no test files]
